### PR TITLE
IMB-78: Set up GOV.UK Notify content for database upload - Success/Failure

### PR DESCRIPTION
**Pull request by jamiecarterHO**
## What?

- Added configurable support for a 'notifications' model and client. This is done in the same way as the DB client and model.
- Added client initialisation and model of methods for GovUK Notify service.
- Added unit tests for above model methods.
- Altered the location of certain data arrays so that they are in scope to be used by the catch block and notify methods.
- Variable scope and error handling is improved (more errors will trigger a notification).
- Updated README

## Why?

[IMB-78](https://collaboration.homeoffice.gov.uk/jira/browse/IMB-78): As an operations/caseworker who has uploaded a CSV list of UANs and DOBs to the HOF IMA form for processing into their database, I would like to receive a notification email of that processing's success or failure so that I know that the IMA form will validate the up to date list of onboarded applicants in Atlas.

## How?

- Added the notifications-node-client and an initialising js file that will export the client
- Added a notify model class with methods passed a 'jobReport' object to formulate a dynamic email request to Notify service.
- Added templates that consume the above request in the IMA service in notify to test.
- Added use of Node.js 'pipeline' instead of stream.pipe as pipeline can is promisified and can be awaited - this has made the main catch block pick up more errors from the script that will trigger an error email.

## Testing

Unit Test assertions added for Notify model functions.
Script and email receipt tested manually.

### CODE REVIEW

File: lib/file-vault-utils.js [hunk 1]
```diff
@@ -31,7 +31,7 @@ const auth = async () =>
       .catch(error => {
         // eslint-disable-next-line no-console
         logger.log('error', 'Keycloak authentication error');
-        reject(error);
+        reject(new Error('Keycloak authentication error', { cause: error }));
```
Comment (sulthan-ahmed) [inline lib/file-vault-utils.js:34]: just minor point, you started with Error in each log above, is it worth doing the same here to be consistent?

File: notifications/models/govuk-notify-model.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      fileUploadTime,
+      jobEndedTime,
+      invalidRecords
+    } = jobReport;
+
+    const emailProps = {
+      csv_uploaded_datetime: this.getDateAndTimeString(fileUploadTime),
+      cepr_update_datetime: this.getDateAndTimeString(jobEndedTime),
+      has_invalid_records: (invalidRecords.length ? 'yes' : 'no'),
+      link_to_file: ''
```
Comment (sulthan-ahmed) [inline notifications/models/govuk-notify-model.js:25]: You have to set `link_to_file` to empty string? Can you just not set it or is it to help with readability and clarity for setting up the emailProps?

File: notifications/models/govuk-notify-model.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      emailProps.records_count = recordsCount;
+      templateId = successTemplateId;
+    } else {
+      emailProps.failure_message = errorMessage;
+      templateId = failureTemplateId;
+    }
+
+    if (invalidRecords.length) {
+      const invalidRecordsCsv = Buffer.from(this.writeInvalidRecordsToCsv(invalidRecords), 'utf8');
+      const fileNameDate = this.getDateAndTimeString(fileUploadTime).replace(/[\/\s:]/g, '-');
```
Comment (sulthan-ahmed) [inline notifications/models/govuk-notify-model.js:38]: no biggie, just an alternative more clearer to me

```
const moment = require('moment');

// Assuming fileUploadTime is a string representing a date and time
const formattedDateTime = moment(fileUploadTime).format('YYYY-MM-DD-HH-mm-ss');

console.log(formattedDateTime);
```

File: notifications/models/govuk-notify-model.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        filename: `invalid-cepr-records-uploaded-${fileNameDate}.csv`
+      });
+    }
+
+    try {
+      return await client.sendEmail(templateId, caseworkerEmail, {
+        personalisation: emailProps
+      });
+    } catch (error) {
+      logger.log('error', 'Error sending notifications:', error);
```
Comment (sulthan-ahmed) [inline notifications/models/govuk-notify-model.js]:
```suggestion
      logger.log('error', 'Error sending notifications email:', error);
```

File: notifications/models/govuk-notify-model.js [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    }
+  }
+
+  createCsvHeadersFromFirstRecord(invalidRecord) {
+    const { record } = invalidRecord;
+    return `${Object.keys(record).join(',')},Reasons for invalidity`;
```
Comment (sulthan-ahmed) [inline notifications/models/govuk-notify-model.js:55]:
```suggestion
    return `${Object.keys(record).join(',')},Reasons for invalid record`;
```
invalidity is just a bit wordy and might not be specific enough. Also should you mention this is related to the CSV?

File: notifications/models/govuk-notify-model.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  writeInvalidRecordsToCsv(invalidRecords) {
+    const dynamicHeaders = this.createCsvHeadersFromFirstRecord(invalidRecords[0]);
+    let csvString = `${dynamicHeaders}\r\n`;
+
+    for (let i = 0; i < invalidRecords.length; i++) {
+      const { reasons } = invalidRecords[i];
+      const stringSizeBytes = Buffer.byteLength(csvString, 'utf8');
+
+      if (stringSizeBytes >= 1999900) {
+        logger.log('warn', 'WARNING: invalid records may exceed max Notify attachment size of 2MB');
```
Comment (sulthan-ahmed) [inline notifications/models/govuk-notify-model.js:67]: Why do you say may exceed?
Is there no specific answer?

Comment (jamiecarterHO) [reply notifications/models/govuk-notify-model.js:67]: I've said 'may' exceed because the actual limit is 2mb whereas the check (and limit to stop adding invalid records to the CSV attached to the notification) is at 100b below this. I was concerned that stopping additions at the actual limit would still cause the file upload in Notify to fail, but It is possible that the list of invalids is between the check limit and the actual limit.

So while it is very unlikely that the list of invalids is in this small margin it is technically not specific. I can improve this message though if you think it is not so useful.

File: lib/file-vault-utils.js (hunk 1 above)
Comment (jamiecarterHO) [reply lib/file-vault-utils.js:34]: The error above is for the Winston logger to colourise red whereas the new error is used for the Notify email and has pretty good context around it already in the template.

File: notifications/models/govuk-notify-model.js (hunk 2 above)
Comment (jamiecarterHO) [reply notifications/models/govuk-notify-model.js:25]: This is because of a quirk of Notify where you can have conditional text based on a personalisation variable but not include another the variable itself within conditional text. 

If set as an empty string by default it will display nothing in the template, and if it is replaced with an actual link it will display this instead. It does help with clarity but the main reason is to avoid creating two new templates for situations where there is a link to display.

File: notifications/models/govuk-notify-model.js (hunk 3 above)
Comment (jamiecarterHO) [reply notifications/models/govuk-notify-model.js:38]: Yes I thought of this, but [moment is probably at end of life](https://momentjs.com/docs/#/-project-status/) so I didn't want to include it in a new project.

File: notifications/models/govuk-notify-model.js (hunk 5 above)
Comment (jamiecarterHO) [reply notifications/models/govuk-notify-model.js:55]: This is one of the column headers in the invalid records CSV e.g. 

```
cepr,dob,dtr,Reasons for invalidity
12345,01/01/2000,Yes,Badly formatted cepr value
```

In my opinion it is grammatically correct, but I'll happily update if you think its a bit archaic?

Comment (sulthan-ahmed) [reply notifications/models/govuk-notify-model.js:55]: it's more about readability and cognitive load for a developer.  It's fine from a grammar perspective.  It's really no biggie so feel free to keep it 🙂

File: test/unit/pgp-model.test.js [hunk 7]
```diff
@@ -34,11 +34,11 @@ describe('The database model method getLatestUrl()', () => {
 
   test('returns an error if failing', async () => {
     client = {
-      one: jest.fn().mockRejectedValue(new Error('Url retrieval failed'))
+      one: jest.fn().mockRejectedValue(new Error('Error retrieving data URL'))
     };
 
     return db.getLatestUrl(client).catch(error => {
-      expect(error.message).toEqual('Url retrieval failed');
+      expect(error.message).toEqual('Error retrieving data URL');
```
Comment (sulthan-ahmed) [inline test/unit/pgp-model.test.js:41]: Error at the beginning is good 👍

File: README.md [hunk 8]
```diff
@@ -40,7 +40,7 @@ Some [documentation of the high level design process for the lookup replacer](ht
 
 ## Configuration
 
-The `config.js` file transforms environment variables into an exported object that can be referred to throughout the app.
+The root level `config.js` file transforms environment variables into an exported object that can be referred to throughout the app.
```
Comment (jamiecarterHO) [inline README.md:44]: Differentiate from service level config file.

File: README.md [hunk 9]
```diff
@@ -75,7 +79,9 @@ module.exports = {
...
 
-If the `validateRecord` property is defined as a function with one `record` argument the script will use that function to validate records row by row. The function must return at least an object with a bool type `valid` property. e.g. `{ valid: true }`. Records returning `{ valid: true }` from this function are added to the `records` array, those that return `{ valid: false }` are added to the `invalidRecords` array. Check existing implementation for examples.
+If the `validateRecord` property is defined as a function with one `record` argument the script will use that function to validate records row by row. The function must return at least an object with at least the `valid` property. Records returning `{ valid: true }` from this function are added to the `records` array, those that return `{ valid: false }` are added to the `invalidRecords` array. Check existing implementation for examples.
+
```
Comment (jamiecarterHO) [inline README.md:83]: Give link to IMA file here to help

File: README.md [hunk 10]
```diff
@@ -85,6 +91,14 @@ This script must authenticate with the service's Keycloak realm and receive a be
 
 It is possible to use a data source that is not protected by Keycloak as long as it either a) supports the same auth protocols as Keycloak in which case environment variables can be substituted like for like, or b) if your data source does not require auth at all. In this case if no Keycloak token URL is provided to the environment the script will not attempt to auth and will pass `undefined` into the Axios data request config - meaning no auth headers are added to the outgoing request to the data URL.
 
+### Notifications
+
+Similar to the Database selection, The Notifications service used can be configured to client and model.
```
Comment (jamiecarterHO) [inline README.md]: lowercase!

File: README.md [hunk 11]
```diff
@@ -85,6 +91,14 @@ This script must authenticate with the service's Keycloak realm and receive a be
...
+
+A notifications client and model is imported in `run.js`. The model used is specified by environment variables `NOTIFICATIONS_CLIENT` and `NOTIFICATIONS_MODEL`. This variable is required. The `govuk-notify` client and `govuk-notify-model` model use the [notifications-node-client](https://docs.notifications.service.gov.uk/node.html) module.
+
+If adding a new notifications client or model use the existing methods and implementation as a guide.
```
Comment (jamiecarterHO) [inline README.md]: add link here too

**Review by sulthan-ahmed — APPROVED** [review]
Great work as always man 👍
